### PR TITLE
Kwd tplg updates

### PR DESCRIPTION
**Pull request by keyonjie**
add bytes controls for both kpb and detector components.

### CODE REVIEW

File: tools/topology/sof/pipe-kfbm-capture.m4 [hunk 1]
```diff
@@ -12,11 +12,37 @@ include(`buffer.m4')
...
 
 dnl Soun Trigger Stream Name)
 define(`N_STS', `Sound Trigger Capture '$1)
 
+#
+# Controls
+#
+
+# kpbm initial parameters
+CONTROLBYTES_PRIV(KPB_priv,
```
Comment (lgirdwood) [inline tools/topology/sof/pipe-kfbm-capture.m4]: what are these ? what structure, are they real coefficients or test data ???

File: tools/topology/m4/detect_coef.m4 [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# initial coe params for detector.
+CONTROLBYTES_PRIV(DETECTOR_priv,
```
Comment (lgirdwood) [inline tools/topology/m4/detect_coef.m4]: structure name ?, real or test data, for which component ? Intel KPD or KPD from other vendors ?

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 1 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: @lgirdwood we are aligned to use private kcontrol bytes for kpb/Intel KPD initialization(the corresponding structs are sof_kpb_config and sof_detect_test_config), maybe I should add these info into comment messages.

File: tools/topology/m4/detect_coef.m4 (hunk 2 above)
Comment (keyonjie) [reply tools/topology/m4/detect_coef.m4]: struct sof_detect_test_config, it is real data for Intel KPD.

**Comment by keyonjie** [comment]
@lgirdwood updated, please help review.

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 1 above)
Comment (lgirdwood) [reply tools/topology/sof/pipe-kfbm-capture.m4]: yes please as other users will use this as an **example**

File: tools/topology/m4/detect_coef.m4 (hunk 2 above)
Comment (lgirdwood) [reply tools/topology/m4/detect_coef.m4]: again, lets document all this in comments.

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 1 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: done.

File: tools/topology/m4/detect_coef.m4 (hunk 2 above)
Comment (keyonjie) [reply tools/topology/m4/detect_coef.m4]: done.

**Comment by keyonjie** [comment]
Hi @mrajwa @tlauda can you help review this please, we need at least one more approval before this can be merged.

File: tools/topology/m4/kpbm.m4 [hunk 3]
```diff
@@ -5,7 +5,7 @@ dnl Define macro for Key Phrase Buffer Manager(kpbm) widget
 dnl kpbm Name)
 define(`N_KPBM', `KPBM'$2`.'$1)
 
-dnl W_KPBM(name, format, periods_sink, periods_source, pipeline_id)
+dnl W_KPBM(name, format, periods_sink, periods_source, pipeline_id, kcontrols_list)
```
Comment (mrajwa) [inline tools/topology/m4/kpbm.m4]: what is this new parameter *kcontrols_list* used for?

File: tools/topology/sof/pipe-kfbm-capture.m4 [hunk 4]
```diff
@@ -26,7 +52,7 @@ define(`N_STS', `Sound Trigger Capture '$1)
 W_PCM_CAPTURE(PCM_ID, Sound Trigger Capture, 0, 2, 2)
 
 # "KPBM" has 2 source and 2 sink periods
-W_KPBM(0, PIPELINE_FORMAT, 2, 2, PIPELINE_ID)
+W_KPBM(0, PIPELINE_FORMAT, 2, 2, PIPELINE_ID, LIST(`             ', "KPB"))
```
Comment (mrajwa) [inline tools/topology/sof/pipe-kfbm-capture.m4]: What is this list for an what do you mean by *"KPBM" has 2 source and 2 sink periods*. KPB has 2 sinks, namely selector and host but as for sources I thought we only have one? We also use only one period, so please clarify it.

Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: > What is this list for an what do you mean by _"KPBM" has 2 source and 2 sink periods_. KPB has

the list is alsa mixer(kcontrol) list we will create for this kpbm widget in driver side.

> 2 sinks, namely selector and host but as for sources I thought we only have one? We also use only one period, so please clarify it.

Here it means we have both 2 periods for source(connected with dmic) and sink(connected with host).

File: tools/topology/sof/pipe-kfbm-capture.m4 [hunk 5]
```diff
@@ -12,11 +12,37 @@ include(`buffer.m4')
...
+`       bytes "0x53,0x4f,0x46,0x00,0x00,0x00,0x00,0x00,'
+`       0x18,0x00,0x00,0x00,0x00,0x10,0x00,0x03,'
+`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
+`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
+`       0x02,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
+`       0x80,0x3e,0x00,0x00,0x10,0x00,0x00,0x00"'
+)
+
+# Selector Bytes control with max value of 255
```
Comment (tlauda) [inline tools/topology/sof/pipe-kfbm-capture.m4]: KPB bytes control?

File: tools/topology/sof/pipe-detect.m4 [hunk 6]
```diff
@@ -40,12 +40,25 @@ C_CONTROLBYTES(SELECTOR, PIPELINE_ID,
 	CONTROLBYTES_MAX(, 304),
 	,
 	SELECTOR_priv)
+
+# Detector initial parameters for Intel KPD.
+include(`detect_test_coef.m4')
+
+# EQ Bytes control with max value of 255
```
Comment (tlauda) [inline tools/topology/sof/pipe-detect.m4]: Detector bytes control?

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 4 above)
Comment (mrajwa) [reply tools/topology/sof/pipe-kfbm-capture.m4]: @keyonjie, could you please say what role will this kcontrol play on FW side?
You send some initial values like *bytes "0x53,0x4f,...* how can one know what they initialize, for what purpose and why is it needed in firmware?

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 5 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: good, let me change, thanks.

File: tools/topology/sof/pipe-detect.m4 (hunk 6 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-detect.m4]: yep, some copy-paste typo here, thanks.

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 4 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: As stated in the comment message(and also I talked to you offline yesterday), it is kpb initial parameters, aligned with struct sof_kpb_config, the first 32Bytes is header as what we have already used for EQ.

**Comment by keyonjie** [comment]
comments addressed, @tlauda @mrajwa

**Comment by tlauda** [comment]
@zrombel Could we rerun the CI?

**Comment by keyonjie** [comment]
looks the CI fail to work, @zrombel?

**Comment by keyonjie** [comment]
@tlauda @lgirdwood could either of you help to merge this if no more changes needed, then it can go to validation soon.

**Comment by tlauda** [comment]
@mrajwa Do you accept the changes?

**Pull request merged**